x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ClientHelper.java

```diff
@@ -77,6 +84,85 @@ public static void assertNoAuthorizationHeader(Map<String, String> headers) {
         }
     }
 
+    /**
+     * In addition to {@link #filterSecurityHeaders}, also check the version of Authentication objects
+     * and rewrite them using minNodeVersion so that they are safe to be persisted as index data
+     * and loaded by all nodes in the cluster.
+     */
+    public static Map<String, String> getPersistableSafeSecurityHeadersForVersion(ThreadContext threadContext, Version minNodeVersion) {
```

Comment:
I think this is trappy.

It assumes that the caller knows what the right version is, which means we're expecting every caller to have the cluster service and get the minimum node version, with the possible exception of accessing a remote cluster.

That means that over time we will end up with callers who just pass in `Version.CURRENT` because they don't understand the implications.

I think we should pass in the `ClusterState` instead.

Reply:
Good point, thanks! I have updated the PR accordingly.